### PR TITLE
[SFSRAP-34] Generate hard-coded sqale file for analyzer

### PR DESCRIPTION
**Pull request by duncanpMS**
TODO: add command line option to allow user to specify the sqale file

### CODE REVIEW

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 1]
```diff
@@ -97,18 +98,24 @@ public bool Generate(NuGetReference analyzeRef, string language, string sqaleFil
...
+
+                if (analyzers.Any())
                 {
+                    definition.RulesFilePath = GenerateRulesFile(analyzers, baseDirectory);
+
+                    if (definition.SqaleFilePath == null)
+                    {
+                        definition.SqaleFilePath = GenerateFixedSqaleFile(analyzers, baseDirectory);
+                    }
+
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: Both the rule generation and sqale generation require the list of analyzers to traverse. I've refactored the old "TryGenerateRulesFile" method to separate out the analyzer discovery into a separate method.

File: RoslynPluginGenerator/AnalyzerPluginGenerator.cs [hunk 2]
```diff
@@ -142,48 +149,72 @@ private bool PackageRequiresLicenseAcceptance(IPackage package)
...
+        /// <returns>The full path to the generated file</returns>
+        private string GenerateRulesFile(IEnumerable<DiagnosticAnalyzer> analyzers, string baseDirectory)
+        {
+            this.logger.LogInfo(UIResources.APG_GeneratingRules);
+
+            Debug.Assert(analyzers.Any(), "Expecting at least one analyzer");
+
+            string rulesFilePath = Path.Combine(baseDirectory, "rules.xml");
+
+            RuleGenerator ruleGen = new RuleGenerator(this.logger);
```
Comment (duncanpMS) [inline RoslynPluginGenerator/AnalyzerPluginGenerator.cs]: Same as the old lines 170 onwards.

File: RoslynPluginGenerator/HardcodedConstantSqaleGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SqaleGenerator.cs" company="SonarSource SA and Microsoft Corporation">
```
Comment (duncanpMS) [inline RoslynPluginGenerator/HardcodedConstantSqaleGenerator.cs]: File name in copyright header is wrong: I'll update it.

File: RoslynPluginGenerator/HardcodedConstantSqaleGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+using SonarLint.XmlDescriptor;
+using SonarQube.Plugins.Common;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace SonarQube.Plugins.Roslyn
+{
+    /// <summary>
+    /// Generates a simple SQALE description using a constant value for each rule
```
Comment (duncanpMS) [inline RoslynPluginGenerator/HardcodedConstantSqaleGenerator.cs]: This class uses the data classes defined in the SonarLint assembly to build a sqale model. These classes are already XML-serializable, so the file is created just by using the XMLSerializer against the model (this is what the C# plugin does).

File: PluginGenerator/PluginBuilder.cs [hunk 5]
```diff
@@ -275,6 +276,18 @@ protected string FindPluginKey()
             return pluginKey;
         }
 
+        protected string FindPackageName()
```
Comment (duncanpMS) [inline PluginGenerator/PluginBuilder.cs]: Previously subclasses didn't have a way to determine the package name (should probably be called "PackageNamespace"). The rules plugin now needs this info as it uses the package namespace to make the names of the rules and sqale files unique.

File: PluginGenerator/RulesPluginBuilder.cs [hunk 6]
```diff
@@ -89,7 +89,11 @@ protected override void ApplyConfiguration(string baseWorkingDirectory)
         {
             this.ValidateRulesConfiguration();
 
-            string uniqueId = Guid.NewGuid().ToString();
+            string packageName = this.FindPackageName(); ;
+            Debug.Assert(!string.IsNullOrWhiteSpace(packageName), "Package name should be set");
+
+            string resourceRoot = packageName.Replace(".", "/") + "/resources/";
+            string absoluteResourceRoot = "/" + resourceRoot;
```
Comment (duncanpMS) [inline PluginGenerator/RulesPluginBuilder.cs]: The rule and sqale xml files used to made unique using guids, which was ugly and harder to work with. They are now qualified using the package namespace.

File: Tests/IntegrationTests/Roslyn/Resources/PluginInspector.java [hunk 7]
```diff
@@ -51,7 +54,8 @@ public static void main(String[] args) throws Exception {
 
         if (args.length != 1 && args.length != 2){
             Log("Expecting at least one argument: the full path to the jar file.\r\nOptionally, the name of the file to be created may also be supplied.");
-			System.exit(ErrorCode);
+            //          jarPath = "C:\\SonarQube\\sonarqube-5.2\\extensions\\plugins\\analyzer1pkgid1-plugin-1.0.2.jar";
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/Resources/PluginInspector.java]: Used when debugging the tests - not required.

File: Tests/IntegrationTests/Roslyn/Resources/RuntimeDependencies.pom [hunk 8]
```diff
@@ -16,6 +16,12 @@
     </dependency>
     
     <dependency>
+      <groupId>org.codehaus.sonar.sslr-squid-bridge</groupId>
+      <artifactId>sslr-squid-bridge</artifactId>
+      <version>2.6</version>
+    </dependency>
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/Resources/RuntimeDependencies.pom]: Jars that embed sqale files need this dependency to execute (contains the SqaleXmlLoader class)

File: Tests/IntegrationTests/Roslyn/Resources/PluginInspector.java [hunk 9]
```diff
@@ -63,21 +67,72 @@ public static void main(String[] args) throws Exception {
...
 
-        Document doc = ProcessJar(jarPath);
+        String xmlFilePath = args.length == 2 ? args[1] : jarPath + ".dump.xml";
+
+        Integer exitCode = CreateAndInvokeInspector(jarPath, xmlFilePath);
+
+        System.exit(exitCode.intValue());
+    }
+
+    public static Integer CreateAndInvokeInspector(String jarPath, String xmlFilePath) throws MalformedURLException, ClassNotFoundException, NoSuchMethodException, InvocationTargetException, IllegalAccessException {
```
Comment (duncanpMS) [inline Tests/IntegrationTests/Roslyn/Resources/PluginInspector.java]: The plugin inspector is Java code that instantiates a plugin, creates the extensions and dumps information about the extensions to an XML file. I haven't changed the XML format. However, I needed to change the way the inspector loaded plugins due to differences in the way the platform RulesXmlLoader and SqaleXmlLoader classes locate the resource files to load.
Previously, the inspector could use a simple UrlClassLoader, which was sufficient to locate the rules xml file. Unfortunately, we need something more complicated to correctly locate the sqale file resource. The comments in the class provide more info: we can discuss offline if necessary.

**Comment by bgavrilMS** [comment]
LGTM

**Pull request merged**